src/main/java/com/vadmack/mongodbtest/service/UserService.java

```diff
@@ -0,0 +1,116 @@
+import com.vadmack.mongodbtest.util.SortService;
+import lombok.RequiredArgsConstructor;
+import org.jetbrains.annotations.Nullable;
+import org.modelmapper.ModelMapper;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Sort;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Service
+@RequiredArgsConstructor
+public class UserService {
+
+    private final UserRepository repository;
+    private final ModelMapper modelMapper = new ModelMapper();
+    private final SequenceGeneratorService sequenceGeneratorService;
+    private final PageableService pageableService;
+    private final SortService sortService;
+    private final ProjectService projectService;
+    private final PasswordEncoder passwordEncoder;
+
+    public List<UserDto> findList(
+            @Nullable String usernamePart,
+            @Nullable Integer pageNumber,
+            @Nullable Integer pageSize,
+            @Nullable String[] sortBy
+    ) {
+        if (usernamePart == null) {
+            usernamePart = "";
+        }
+        pageableService.validateParams(pageNumber, pageSize);
+        Sort sort = sortService.createSort(sortBy);
+        if (pageNumber != null && pageSize != null) {
+            if (sortBy == null) {
+                sortBy = new String[1];
+                sortBy[0] = "id:0";
+            }
+            return repository.findAllByUsernameLikeIgnoreCase(
+                    usernamePart,
+                    PageRequest.of(pageNumber, pageSize, sort))
+                    .stream().map(this::entityToDto)
+                    .collect(Collectors.toList());
+        } else {
+            return repository.findAllByUsernameLikeIgnoreCase(usernamePart, sort)
+                    .stream().map(this::entityToDto)
+                    .collect(Collectors.toList());
+        }
+    }
+
+    public UserDto findById(Long id) {
+        return entityToDto(getById(id));
+    }
+
+    public void create(UserNoIdDto userNoIdDto) {
+        validateAuthorities(userNoIdDto);
+        User user = dtoToEntity(userNoIdDto);
+        user.setId(sequenceGeneratorService.generateSequence(User.SEQUENCE_NAME));
+        user.setPassword(passwordEncoder.encode(userNoIdDto.getPassword()));
+        repository.save(user);
+    }
+
+    public void update(Long id, UserNoIdDto userNoIdDto) {
+        validateAuthorities(userNoIdDto);
+        getById(id);
+        User user = dtoToEntity(userNoIdDto);
+        user.setId(id);
+        user.setPassword(passwordEncoder.encode(userNoIdDto.getPassword()));
+        repository.save(user);
+    }
+
+    public void delete(Long id) {
+        User user = getById(id);
+        List<Project> projects = projectService.findAllByOwnerId(id);
+        projects.forEach(project -> projectService.delete(project.getId()));
```

Comment:
What happens if your code will be interrupted  during removing fifth of 10 projects for example?